Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1814 @@
+
+  GS_RANGE_CHECK(aRange, c);
+
+  for (i = aRange.location; i < e; i++)
+    aBuffer[j++] = (*get)(self, oaiSel, i);
+}
+ 
+// Key value coding support
+- (void) setValue: (id)value forKey: (NSString*)key
+{
+  NSUInteger    i;
+  NSUInteger	count = [self count];
+  volatile id	object = nil;
+
+  for (i = 0; i < count; i++)
+    {
+      object = [self objectAtIndex: i];
+      [object setValue: value
+		forKey: key];
+    }
+}
+
+- (id) valueForKey: (NSString*)key
+{
+  NSEnumerator *e = [self objectEnumerator];
+  id object = nil;
+  NSMutableSet *results = [NSMutableSet setWithCapacity: [self count]];
+
+  while ((object = [e nextObject]) != nil)
+    {
+      id result = [object valueForKey: key];
+
+      if (result == nil)
+        continue;
+
+      [results addObject: result];
+    }
+  return results;
+}
+
+// Key-Value Observing Support
+/*
+- addObserver:forKeyPath:options:context:
+- removeObserver:forKeyPath:
+- removeObserver:forKeyPath:context:
+*/
+
+- (NSUInteger)_countForObject: (id)object  // required override...
+{
+  return 1;
+}
+
+// Comparing Sets
+- (BOOL) isEqualToOrderedSet: (NSOrderedSet *)aSet
+{
+  if ([self count] == 0 &&
+      [aSet count] == 0)
+    return YES;
+  
+  if (self == aSet)
+    return YES;
+  
+  if ([self count] != [aSet count])
+    return NO;
+
+  // if they are equal, then this set will be a subset of aSet.
+  return [self isSubsetOfOrderedSet: aSet];
+}
+
+- (BOOL) isEqual: (id)other
+{
+  if ([other isKindOfClass: [NSOrderedSet class]])
+    return [self isEqualToOrderedSet: other];
+  return NO;
+}
+
+// Set operations
+- (BOOL) intersectsOrderedSet: (NSOrderedSet *)otherSet
+{
+  id	o = nil, e = nil;
+
+  // -1. If this set is empty, this method should return NO.
+  if ([self count] == 0)
+    return NO;
+
+  // 0. Loop for all members in otherSet
+  e = [otherSet objectEnumerator];
+  while ((o = [e nextObject])) // 1. pick a member from otherSet.
+    {
+      if ([self containsObject: o])    // 2. check the member is in this set(self).
+        return YES;
+    }
+  return NO;
+}
+
+- (BOOL) intersectsSet: (NSSet *)otherSet
+{
+  id	o = nil, e = nil;
+
+  // -1. If this set is empty, this method should return NO.
+  if ([self count] == 0)
+    return NO;
+
+  // 0. Loop for all members in otherSet
+  e = [otherSet objectEnumerator];
+  while ((o = [e nextObject])) // 1. pick a member from otherSet.
+    {
+      if ([self containsObject: o])    // 2. check the member is in this set(self).
+        return YES;
+    }
+  return NO;
+}
+
+- (BOOL) isSubsetOfOrderedSet: (NSOrderedSet *)otherSet
+{
+  id    f = nil;
+  NSUInteger s = 0, l = [self count], i = 0;
+  
+  // -1. If this set is empty, this method should return YES.
+  if (l == 0)
+    return YES;
+
+  // If count of set is more than otherSet it's not a subset
+  if (l > [otherSet count])
+    return NO;
+
+  // Find the first object's index in otherset, if not found
+  // it's not a subset...
+  f = [self firstObject];
```

Comment:
As I understand the logic here, the subset needs to have the same objects as the super set and also in the same order but the super set may have a few extra objects in between. This is easy to implement with two enumerators. The one for self gets advanced when ever the current two objects are the same. The other enumerator gets increased for each step. If self get exhausted we return YES.  If the other is exhausted first we return NO.

Reply:
Implemented as you suggest.